### ISSUE TITLE
[Bug Report] Expected URL that uses protocol HTTPS but received one that uses protocol 'http'

### ISSUE THREAD

**Issue by fnbk** [Azure/azure-iot-sdk-java#914]
The Device Client API `HttpsConnection` Function only supports HTTPS connections and throws an exception when HTTP connections are used. 

https://github.com/Azure/azure-iot-sdk-java/blob/11f1ea95ff47838b9fe42b106ebce3d28e2743cd/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/https/HttpsConnection.java#L60-L99

# Background

We are using the Device Client API in a module that connects to the iotedge daemon to read module-twin settings. 

Based on the edge runtime (edge or edge-on-k8s) a different URL will be provided by default:

**edge**
```
IOTEDGE_WORKLOADURI=unix:///var/run/iotedge/workload.sock
```

**edge-on-k8s**
```
IOTEDGE_WORKLOADURI=http://localhost:35001/
```

So in the regular edge environment (moby-engine) there are no issues, but in the kubernetes environment while a sidecar proxy is used to create connections to the edge daemon, the default configuration (`IOTEDGE_WORKLOADURI=http://localhost:35001/`) leads to the following error:

```
08:28:37,503 WARN  [com.iqs.cq.kc.twinconfig.TwinConfigurationProvider] (ServerService Thread Pool -- 71) post init, twinMonitor=null
08:28:37,522 INFO  [com.microsoft.azure.sdk.iot.device.ModuleClient] (ServerService Thread Pool -- 71) Creating module client from environment with protocol MQTT...
08:28:37,522 INFO  [com.microsoft.azure.sdk.iot.device.ModuleClient] (ServerService Thread Pool -- 71) No connection string was configured for this module, so it will get its credentials from the edgelet
08:28:37,577 INFO  [stdout] (ServerService Thread Pool -- 71) Exception
08:28:37,577 ERROR [stderr] (ServerService Thread Pool -- 71) java.lang.IllegalArgumentException: Expected URL that uses protocol HTTPS but received one that uses protocol 'http'.
08:28:37,577 ERROR [stderr] (ServerService Thread Pool -- 71)
08:28:37,578 ERROR [stderr] (ServerService Thread Pool -- 71)   at deployment.iqs-cq-kctwin-bundle-1.0.0-SNAPSHOT.ear.com.microsoft.azure.sdk.iot-iot-device-client-1.25.0.jar//com.microsoft.azure.sdk.iot.device.transport.https.HttpsConnection.<init>(HttpsConnection.java:79)
08:28:37,578 ERROR [stderr] (ServerService Thread Pool -- 71)   at deployment.iqs-cq-kctwin-bundle-1.0.0-SNAPSHOT.ear.com.microsoft.azure.sdk.iot-iot-device-client-1.25.0.jar//com.microsoft.azure.sdk.iot.device.transport.https.HttpsRequest.send(HttpsRequest.java:102)
08:28:37,578 ERROR [stderr] (ServerService Thread Pool -- 71)   at deployment.iqs-cq-kctwin-bundle-1.0.0-SNAPSHOT.ear.com.microsoft.azure.sdk.iot-iot-device-client-1.25.0.jar//com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:204)
08:28:37,578 ERROR [stderr] (ServerService Thread Pool -- 71)   at deployment.iqs-cq-kctwin-bundle-1.0.0-SNAPSHOT.ear.com.microsoft.azure.sdk.iot-iot-device-client-1.25.0.jar//com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:82)
```

This issue could be quickly resolved if the `HttpsConnection` function would also support the HTTP protocol. By removing the [guard statement](https://github.com/Azure/azure-iot-sdk-java/blob/11f1ea95ff47838b9fe42b106ebce3d28e2743cd/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/https/HttpsConnection.java#L73), we managed to fix this issue:

```
17:30:50,733 WARN  [com.iqs.cq.kc.twinconfig.TwinConfigurationProvider] (ServerService Thread Pool -- 69) post init, twinMonitor=null
17:30:50,746 INFO  [com.microsoft.azure.sdk.iot.device.ModuleClient] (ServerService Thread Pool -- 69) Creating module client from environment with protocol MQTT...
17:30:50,750 INFO  [com.microsoft.azure.sdk.iot.device.ModuleClient] (ServerService Thread Pool -- 69) No connection string was configured for this module, so it will get its credentials from the edgelet
17:30:52,174 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) The connection to the IoT Hub has been established
17:30:52,175 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Updating transport status to new status CONNECTED with reason CONNECTION_OK
17:30:52,175 INFO  [stdout] (ServerService Thread Pool -- 69)
17:30:52,175 INFO  [stdout] (ServerService Thread Pool -- 69) CONNECTION STATUS UPDATE: CONNECTED
17:30:52,176 INFO  [stdout] (ServerService Thread Pool -- 69) CONNECTION STATUS REASON: CONNECTION_OK
17:30:52,176 INFO  [stdout] (ServerService Thread Pool -- 69) CONNECTION STATUS THROWABLE: null
17:30:52,177 INFO  [stdout] (ServerService Thread Pool -- 69)
17:30:52,182 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Client connection opened successfully
17:30:52,193 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Message was queued to be sent later ( Message details: Correlation Id [f66f924a-7c77-48d0-9597-ea953fe62cb6] Message Id [6fc1ccc8-142e-47fe-be79-a39e6e47ea70] Device Operation Type [DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST] )
17:30:52,196 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Message was queued to be sent later ( Message details: Correlation Id [3b480997-37df-409f-a111-e527591dd52f] Message Id [2e061470-2fc8-4145-9b66-246176fade4a] Request Id [0] Device Operation Type [DEVICE_OPERATION_TWIN_GET_REQUEST] )
17:30:52,197 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Message was queued to be sent later ( Message details: Correlation Id [1c984e96-839c-4872-9082-36f16e920086] Message Id [2a9b57bb-67d5-456b-bf4f-3a8b4844a710] Device Operation Type [DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST] )
17:30:52,206 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Message was queued to be sent later ( Message details: Correlation Id [1b71d248-504b-4121-9ae7-fdafa8ecebd4] Message Id [7d40a223-5b0f-44d0-9467-9484a10ac492] Request Id [1] Device Operation Type [DEVICE_OPERATION_TWIN_GET_REQUEST] )
17:30:52,196 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (azure-iot-sdk-IotHubSendTask) Sending message ( Message details: Correlation Id [f66f924a-7c77-48d0-9597-ea953fe62cb6] Message Id [6fc1ccc8-142e-47fe-be79-a39e6e47ea70] Device Operation Type [DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST] )
17:30:52,208 INFO  [com.microsoft.azure.sdk.iot.device.transport.IotHubTransport] (ServerService Thread Pool -- 69) Message was queued to be sent later ( Message details: Correlation Id [182092bf-eee9-4db8-9d4a-3e5f018612af] Message Id [50fc2fb7-5e76-408d-91f6-1fa7eeadcbef] Device Operation Type [DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST] )
```

AB#8252716

**Comment by timtay-microsoft** [comment]
In general, we see HTTP connections as security vulnerabilities. In the case of localhost connections, there may be room for it though. Would it be sufficient for your needs if the SDK allowed HTTP connections only when the host is localhost?

**Comment by timtay-microsoft** [comment]
Can you provide the full stacktrace on this exception?
```
08:28:37,577 ERROR [stderr] (ServerService Thread Pool -- 71) java.lang.IllegalArgumentException: Expected URL that uses protocol HTTPS but received one that uses protocol 'http'.
```

It looks like our HttpsHsmClient class is using this HTTPS only client code even when the workload uri is HTTP, which would be a bug, but I can't tell if that is how you are hitting this exception.

https://github.com/Azure/azure-iot-sdk-java/blob/997e9a82abfffd8560207df9a8ae12025e56ecc0/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/hsm/HttpsHsmClient.java#L202

**Comment by timtay-microsoft** [comment]
And for what it's worth, device client versions 1.20.2 and below do not have this check, so you could use those in the meantime if you need a quick fix

**Comment by timtay-microsoft** [comment]
If the callstack is coming from HttpsHsmClient, then could you try building from [this ](https://github.com/Azure/azure-iot-sdk-java/tree/timtay/http)branch to see if it resolves your issue?

**Comment by fnbk** [comment]
Thank you for your quick response. 

The request is definitely coming from the HttpsHsmClient, but I have no detailed stacktrace at hand right now.

In our "iothub on kubernetes" scenario there are only localhost connections at play - in case this poses any security risks.

I will also build and check your branch tomorrow morning (german time) :-)

Good Job.

**Comment by fnbk** [comment]
@timtay-microsoft I built and deployed [your branch](https://github.com/Azure/azure-iot-sdk-java/tree/timtay/http) and it worked - the connection could be created and module-twin settings were fetched.

**Comment by timtay-microsoft** [comment]
I'll get this change merged into master and release soon then. Thanks for confirming!

**Comment by timtay-microsoft** [comment]
This issue has been fixed as of device client version 1.26.0, so I'm closing this issue

**Comment by az-iot-builder-01** [comment]
@timtay-microsoft, @fnbk, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
